addon/types/map_export/brush.py

```diff
@@ -62,16 +76,54 @@ def calc_uv_axes(settings: typing.Any, bm: bmesh.types.BMesh, face: bmesh.types.
     u1, u2, u3 = u_vals
     v1, v2, v3 = v_vals
 
-    tangent = -1 * ((p2 - p1) * (v3 - v1) - (p3 - p1) * (v2 - v1))
-    bitangent = mathutils.Quaternion(face.normal, math.radians(90)) @ tangent
+    uv_side_a = mathutils.Vector((u2 - u1, v2 - v1))
+    uv_side_b = mathutils.Vector((u3 - u1, v3 - v1))
+    tangent = (p2 - p1) * uv_side_b.y - (p3 - p1) * uv_side_a.y
+    side_v = (p3 - p1) * uv_side_a.x - (p2 - p1) * uv_side_b.x
+    determinant = uv_side_a.x * uv_side_b.y - uv_side_b.x * uv_side_a.y
+
+    epsilon = 0.0000001
 
-    # TODO: Calculate scale and offset
+    if abs(determinant) > epsilon:
+        tangent = tangent / determinant
+        side_v = side_v / determinant
+
+    if not settings.allow_skewed_textures:
+        bitangent = mathutils.Quaternion(face.normal, math.radians(90)) @ tangent
+        bitangent.normalize()
+        bitangent = bitangent * side_v.dot(bitangent)
+    else:
+        bitangent = side_v
+
+    # Scale
+    texture_side_length = get_texture_size(obj, face)
+    if texture_side_length > 0:
+        u_scale = tangent.magnitude / texture_side_length
+        v_scale = bitangent.magnitude / texture_side_length
+    else:
+        u_scale = 1
+        v_scale = 1
+
+    # Offset
+    tangent_space_transform = mathutils.Matrix((
+        [tangent.x, bitangent.x, face.normal.x],
+        [tangent.y, bitangent.y, face.normal.y],
+        [tangent.z, bitangent.z, face.normal.z]
+    ))
+    if texture_side_length > 0 and abs(tangent_space_transform.determinant()) > epsilon:
+        tangent_space_transform.invert()
+        t1 = tangent_space_transform @ p1
+        u_offset = (1 - (t1.x - u1) * texture_side_length) % texture_side_length
+        v_offset = (1 - (t1.y - v1) * texture_side_length) % texture_side_length
+    else:
+        u_offset = 0
+        v_offset = 0
 
     tangent.normalize()
     bitangent.normalize()
 
-    u_axis = f'[{tangent[0]} {tangent[1]} {tangent[2]} 0] {settings.texture_scale}'
-    v_axis = f'[{bitangent[0]} {bitangent[1]} {bitangent[2]} 0] {settings.texture_scale}'
+    u_axis = f'[{tangent[0]} {tangent[1]} {tangent[2]} {u_offset}] {u_scale * settings.texture_scale}'
+    v_axis = f'[{bitangent[0]} {bitangent[1]} {bitangent[2]} {v_offset}] {v_scale * settings.texture_scale}'
```

Comment:
I'm not even going to pretend to understand any of this fancy math stuff, but good job!

Reply:
Right yeah this isn't an issue so I'll resolve it.